mark.c

```diff
@@ -280,6 +280,43 @@ GC_API int GC_CALL GC_is_marked(const void *p)
     return (int)mark_bit_from_hdr(hhdr, bit_no); /* 0 or 1 */
 }
 
+GC_API void GC_CALL GC_set_uncollectable(const void *p)
+{
+    struct hblk *h = HBLKPTR(p);
+    hdr * hhdr = HDR(h);
+    word bit_no = MARK_BIT_NO((word)((ptr_t)p - (ptr_t)h), hhdr -> hb_sz);
+
+    if (!uncollectable_bit_from_hdr(hhdr, bit_no)) {
+      set_uncollectable_bit_from_hdr(hhdr, bit_no);
+      INCR_MARKS(hhdr);
+    }
+}
+
+GC_API void GC_CALL GC_clear_uncollectable(const void *p)
+{
+    struct hblk *h = HBLKPTR(p);
+    hdr * hhdr = HDR(h);
+    word bit_no = MARK_BIT_NO((word)((ptr_t)p - (ptr_t)h), hhdr -> hb_sz);
+
+    if (uncollectable_bit_from_hdr(hhdr, bit_no)) {
+      size_t n_marks = hhdr -> hb_n_marks;
+
+      GC_ASSERT(n_marks != 0);
+      clear_uncollectable_bit_from_hdr(hhdr, bit_no);
+      n_marks--;
+      hhdr -> hb_n_marks = n_marks;
```

Comment:
Is there a reason for the "read into n_marks, then decrement it, then write it" dance? If I did something like:

```c
GC_ASSERT(hhdr->hb_n_marks != 0);
clear_uncollectable_...
hhdr->hb_n_marks --;
```

would this still be correct?

Reply:
I've looked into this and realised that I'd confused things here. We don't want to adjust the marks at all when hopping between collectable / uncollectable objects. I've pushed a fix here: c65972b0f6bd09ee6ed4664f1c3f2b4b9786e375